In [21]:
# set seed for reproducibility
set.seed(1009)

_* Read: Dalgaard, 1.2.12 Conditional selection*_

# Example 1 (Practice)

Consider birth names of 2003.   

1. Find P(first name is Evelyn)   

2. P(first name is Evelyn | female)  

3. P(female | first name is Evelyn)  


In [22]:
# read in the data, name the columns, and view the first few rows
yob2003 <- read.csv("yob2003.txt",header=FALSE, stringsAsFactors=FALSE)
names(yob2003) <- c("name","gender","count")
head(yob2003)

name    gender count
1 Emily   F      25685
2 Emma    F      22701
3 Madison F      20197
4 Hannah  F      17628
5 Olivia  F      16142
6 Abigail F      15918

In [23]:
# filter the data
yob2003.Evelyn <- yob2003[yob2003$name == "Evelyn",] # Evelyn only
yob2003.Evelyn.female <- yob2003.Evelyn[yob2003.Evelyn$gender == "F",] # Evelyn and Female
yob2003.female <- yob2003[yob2003$gender == "F",] # Female only

 # 1. P(Evelyn)
sum(yob2003.Evelyn$count)/sum(yob2003$count)

# 2. P(Evelyn | female)
sum(yob2003.Evelyn.female$count)/sum(yob2003.female$count) 

# 3. P(female | Evelyn)
sum(yob2003.Evelyn.female$count)/sum(yob2003.Evelyn$count) 

[1] 0.0009771525

[1] 0.002026999

[1] 0.9967672

# Example 2: Experiment: Deck of poker cards

## a. Let's generate a deck of cards,

In [24]:
# build a deck of cards (seen below)
suits <- c("Diamonds", "Clubs", "Hearts", "Spades")
numbers <- c("Ace", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Jack", "Queen", "King")
(deck <- expand.grid(number=numbers, suit=suits)) #data frame from all combinations
(deck <- paste(deck$number, deck$suit))

number suit    
1  Ace    Diamonds
2  Two    Diamonds
3  Three  Diamonds
4  Four   Diamonds
5  Five   Diamonds
6  Six    Diamonds
7  Seven  Diamonds
8  Eight  Diamonds
9  Nine   Diamonds
10 Ten    Diamonds
11 Jack   Diamonds
12 Queen  Diamonds
13 King   Diamonds
14 Ace    Clubs   
15 Two    Clubs   
16 Three  Clubs   
17 Four   Clubs   
18 Five   Clubs   
19 Six    Clubs   
20 Seven  Clubs   
21 Eight  Clubs   
22 Nine   Clubs   
23 Ten    Clubs   
24 Jack   Clubs   
25 Queen  Clubs   
26 King   Clubs   
27 Ace    Hearts  
28 Two    Hearts  
29 Three  Hearts  
30 Four   Hearts  
31 Five   Hearts  
32 Six    Hearts  
33 Seven  Hearts  
34 Eight  Hearts  
35 Nine   Hearts  
36 Ten    Hearts  
37 Jack   Hearts  
38 Queen  Hearts  
39 King   Hearts  
40 Ace    Spades  
41 Two    Spades  
42 Three  Spades  
43 Four   Spades  
44 Five   Spades  
45 Six    Spades  
46 Seven  Spades  
47 Eight  Spades  
48 Nine   Spades  
49 Ten    Spades  
50 Jack   Spades  
51 Queen  Spades  
52 King   Spades

[1] "Ace Diamonds"   "Two Diamonds"   "Three Diamonds" "Four Diamonds" 
 [5] "Five Diamonds"  "Six Diamonds"   "Seven Diamonds" "Eight Diamonds"
 [9] "Nine Diamonds"  "Ten Diamonds"   "Jack Diamonds"  "Queen Diamonds"
[13] "King Diamonds"  "Ace Clubs"      "Two Clubs"      "Three Clubs"   
[17] "Four Clubs"     "Five Clubs"     "Six Clubs"      "Seven Clubs"   
[21] "Eight Clubs"    "Nine Clubs"     "Ten Clubs"      "Jack Clubs"    
[25] "Queen Clubs"    "King Clubs"     "Ace Hearts"     "Two Hearts"    
[29] "Three Hearts"   "Four Hearts"    "Five Hearts"    "Six Hearts"    
[33] "Seven Hearts"   "Eight Hearts"   "Nine Hearts"    "Ten Hearts"    
[37] "Jack Hearts"    "Queen Hearts"   "King Hearts"    "Ace Spades"    
[41] "Two Spades"     "Three Spades"   "Four Spades"    "Five Spades"   
[45] "Six Spades"     "Seven Spades"   "Eight Spades"   "Nine Spades"   
[49] "Ten Spades"     "Jack Spades"    "Queen Spades"   "King Spades"

## b.  Let's check the probability of getting the first card a Queen card (we know that it's 1/13)

In [25]:
# define the Queen cards
(queens <- paste("Queen", suits))

# find how many cards in the deck are Queens
mean(deck %in% queens)

# result should be 4/52 = 1/13, so let's check
1/13

[1] "Queen Diamonds" "Queen Clubs"    "Queen Hearts"   "Queen Spades"

[1] 0.07692308

[1] 0.07692308

This means it works!!

## c. Now, what is the conditional probability of the second card being a Queen given that the first was a Queen?

If one Queen is already out of the deck, then there are 3 Queens left and 51 total cards left. So, this probability is 3/51. Let’s confirm by listing out all possible outcomes.

To do this, we can use the permutations function from the "gtools" package. For any list of size `n`, this function computes all the different permutations we can get when we select `r` items. 

### NOTE: "permutations"

Here are all the ways we can choose two numbers from a list consisting of 1,2,3:

In [26]:
library(gtools)
permutations(3, 2)

[,1] [,2]
[1,] 1    2   
[2,] 1    3   
[3,] 2    1   
[4,] 2    3   
[5,] 3    1   
[6,] 3    2

Notice that the order matters here: 3,1 is different than 1,3. Also, note that (1,1), (2,2), and (3,3) do not appear because once we pick a number, it can’t appear again.

Optionally, we can add a vector. If you want to see five random seven digit phone numbers out of all possible phone numbers (without repeats), you can type:

In [27]:
# generate all phone numbers without repeating digits
all_phone_numbers <- permutations(10, 7, v = 0:9) #here 10 is the size of the vector

# display 5 examples of phone numbers generated
(n <- nrow(all_phone_numbers))
(index <- sample(n, 5))
all_phone_numbers[index,]

[1] 604800

[1] 254598 138510 115335 495466 536734

[,1] [,2] [,3] [,4] [,5] [,6] [,7]
[1,] 4    1    9    0    6    8    3   
[2,] 2    3    6    9    1    5    4   
[3,] 1    9    2    4    0    7    6   
[4,] 8    1    6    7    9    2    3   
[5,] 8    7    6    9    4    3    1

Instead of using the numbers 1 through 10, the default, it uses what we provided through v: the digits 0 through 9.


### _*Let's go to our earlier example*_

To compute all possible ways we can choose two cards when the order matters, we type:


In [28]:
hands <- permutations(52, 2, v = deck)
head(hands)
nrow(hands) # number of possible ways to choose two cards when order matters

[,1]      [,2]          
[1,] Ace Clubs Ace Diamonds  
[2,] Ace Clubs Ace Hearts    
[3,] Ace Clubs Ace Spades    
[4,] Ace Clubs Eight Clubs   
[5,] Ace Clubs Eight Diamonds
[6,] Ace Clubs Eight Hearts

[1] 2652

This is a matrix with two columns and 2652 rows. With a matrix we can get the first and second cards like this:

In [29]:
first_card <- hands[,1]
second_card <- hands[,2]

Now the cases for which the first hand was a Queen can be computed like this:

In [30]:
(queens <- paste("Queen", suits))
sum(first_card %in% queens)


mean(first_card %in% queens)
4/52 #prob of a queen card

[1] "Queen Diamonds" "Queen Clubs"    "Queen Hearts"   "Queen Spades"

[1] 204

[1] 0.07692308

[1] 0.07692308

To get the conditional probability, we compute what fraction of these have a Queen in the second card

In [31]:
# given that first card is Queen, how many also had second card as Queen
sum(first_card%in%queens & second_card%in%queens) / sum(first_card%in%queens)
3/51

[1] 0.05882353

[1] 0.05882353

which is exactly 3/51, as we had already deduced. Notice that the code above is equivalent to:

In [32]:
mean(first_card%in%queens & second_card%in%queens) / mean(first_card%in%queens)

[1] 0.05882353

# Example 3: Blackjack

In Blackjack if you get an Ace, and a face card or ten in the first draw, it is called a Natural 21 and you win automatically. If we wanted to compute the probability of this happening, we would enumerate the combinations, not the permutations, since the order does not matter.

So to compute the probability of a Natural 21 in Blackjack, we can do this:

In [33]:
# generate Aces
(aces <- paste("Ace", suits))

# generate face cards and tens together
facecard <- c("King", "Queen", "Jack","Ten")
(facecard <- expand.grid(number = facecard, suit = suits))
facecard <- paste(facecard$number, facecard$suit)

# compute number of Natural 21 hands
hands <- combinations(52, 2, v = deck)
mean(hands[,1] %in% aces & hands[,2] %in% facecard)

[1] "Ace Diamonds" "Ace Clubs"    "Ace Hearts"   "Ace Spades"

number suit    
1  King   Diamonds
2  Queen  Diamonds
3  Jack   Diamonds
4  Ten    Diamonds
5  King   Clubs   
6  Queen  Clubs   
7  Jack   Clubs   
8  Ten    Clubs   
9  King   Hearts  
10 Queen  Hearts  
11 Jack   Hearts  
12 Ten    Hearts  
13 King   Spades  
14 Queen  Spades  
15 Jack   Spades  
16 Ten    Spades

[1] 0.04826546

In the last line, we assume the Ace comes first. This is only because we know the way combination enumerates possibilities and it will list this case first. But to be safe, we could have written this and produced the same answer:

In [34]:
# first card ace, second card facecard/ten OR second card ace, first card facecard/ten
mean((hands[,1] %in% aces & hands[,2] %in% facecard) |
       (hands[,2] %in% aces & hands[,1] %in% facecard)) 

[1] 0.04826546

# Bayes' Rule

# Example 3: Disease Testing (this is Fictional)

_Suppose we have a test for the flu that is positive 90% of the time when tested on a flu patient , and is negative 95% of the time when tested on a healthy person . We also know that the flu is affecting about 1% of the population. You go to the doctor and test positive. What is the chance that you truly have the flu?_

$(P(test + | flu) = 0.9)$
$(P(test - | no \;flu) = 0.95)$
$P(flu)=0.01$

You can answer this question directly using Bayes' theorem, but we'll tackle this a bit differently. We'll create a hypothetical population of 100,000 people, and see if we can figure this out.

In [35]:
# 0.01 probability of having flu ('Yes')
flu <- sample(c('No','Yes'), size=100000, replace=TRUE,prob=c(0.99,0.01))
test <- rep(NA, 100000) # create a dummy variable first
# test results for people without flu ('No') and with flu ('Yes')
test[flu=='No'] <- sample(c('Neg','Pos'), size=sum(flu=='No'), replace=TRUE, prob=c(0.95,0.05))
test[flu=='Yes'] <- sample(c('Neg','Pos'), size=sum(flu=='Yes'), replace=TRUE,prob=c(0.1, 0.9))

In the above code we first simulate who has the flu, given on average 1% of the population gets the flu. We then find out whom among those without the flu would test positive, based on $P(test - | no flu) =0.95$. 
Recall that the when considering a conditioning event, the conditioning event is considered the sample space, and so all the laws of probability hold within that space.

In particular:


$P(test + | no \;flu) = 1 - P(test - | no \;flu) = 0.05$


We do a similar computation for the people with flu.

The question we are asking, _what is the chance that you have the flu given that you tested positive_, can then be directly answered as:

In [36]:
mean(flu[test=='Pos']=='Yes')

[1] 0.1597378

Even though the test is pretty good, the chance that we actually have the flu even if we test positive is actually pretty small. This is because the chance of actually getting the flu is pretty small in the first place. However, if we look at how much our _chance of having the flu changed with a positive test_, it is quite large:

In [37]:
mean(flu[test=='Pos']=='Yes')/mean(flu=='Yes')

[1] 15.78437

That is, the knowledge that we tested positive increased our chance of truly having the flu 15-fold! A constant issue in medicine is if we should address the absolute increase in risk (1% to 15%) or the relative risk (15-fold) when deciding on best clinical practice.

If we calculate the probability using Bayes' theorem, we get a very similar result:

P(flu=Yes|test+) = P(test+|flu=Yes)P(flu=Yes)/P(test+)

=P(test+|flu=Yes)P(flu=Yes)/[P(test+|flu=Yes)P(flu=Yes)+P(test+|flu=No)P(flu=No)]

= 0.9 x 0.01 / (0.9 x 0.01 + 0.05 x 0.99)

= 0.1538

# Conditional Distributions 

Text adapted from chapter 25 of Resampling: The New Statistics.

The importance of accurate testing is critical for understanding of disease in a community. We can frame the classic screening for a disease question for a timely example like COVID-19. This is a relevant example because when performing a test for COVID-19, the prevalence of false positives (positive test result when the person does not have COVID-19) and false negatives (negative test result when the person does have COVID-19) is critical for the overall effectiveness of the study. 

Let's say we perform a test for COVID-19 in a population of 100,000. The estimated prevalence of COVID-19 in this population is 5,000 out of the population of 100,000. When a person has COVID-19, 95% of the time the test comes back positive (or a 5% false-negative rate). When a person does not have COVID-19, 5% of the time the test comes back positive for COVID-19 (or a 5% false-positive rate).

(Visual Reference:<https://towardsdatascience.com/interactive-visualizing-covid-19-test-accuracy-4a24b4e1ffbf>)


In [38]:
library(dplyr)
library(ggplot2)
library(knitr)

knitr::opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)


Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




In [39]:
library(tidyverse)

# this function will return the appropriate probabilities
# that you need to answer the questions
# while you don't need to write this yourself, it's encouraged to understandwhat is happening with this code

# first and foremost, make sure you figure out the values to pass to the function

func_testing <- function(num_people, 
                         prop_covid, 
                         posrate_covid, 
                         posrate_nocovid){
    
# make a dataframe by getting uniform random variables 
    # for population and positivity
  
df <- tibble(pop_rv = runif(num_people), 
             pos_rv = runif(num_people))

# set a new variable to be if the person has covid
# base on proportion of covid population

df <- df %>% 
    mutate(covid_yes = if_else(pop_rv <= prop_covid, TRUE, FALSE))

# prob covid, counts all the number of people with covid,
# and divides by whole population
# this value should be close to the prob_covid input

prob_covid <- df %>% 
    summarize(sum(covid_yes)) %>% 
    pull() / num_people

# clearly prob_nocovid is what remains of the probability space
# only two optinos for this example

prob_nocovid <- 1 - prop_covid


# here we add the test positive variable based on the results
# of the random variable for positivity
# note that the result is different if you have covid or don't

df <- df %>% 
    mutate(test_positive = 
               case_when( covid_yes == TRUE & pos_rv <= posrate_covid ~ TRUE, covid_yes == FALSE & pos_rv <= posrate_nocovid ~ TRUE, TRUE ~ FALSE))

# here is the magic of tidyverse
# get back a summarize of the mean positive test for
# the people who have covid, and the people who don't

temp  <- df %>% 
    group_by(covid_yes) %>% 
    summarise(mean_positive = mean(test_positive), .groups = 'keep')

# we pull those values into variables

prob_pos_given_nocovid = temp[1,2] %>% pull() # positive given covid
prob_pos_given_covid = temp[2,2] %>% pull() # positive given no covid

# and we return the result for you as a list

return (list('prob_covid' = prob_covid, 
             'prob_nocovid' = prob_nocovid, 
             'prob_pos_given_covid' = prob_pos_given_covid, 
             'prob_pos_given_nocovid' = prob_pos_given_nocovid)
)
}

-- Attaching packages --------------------------------------- tidyverse 1.3.2 --
v tibble  3.1.8     v purrr   1.0.1
v tidyr   1.3.0     v stringr 1.5.0
v readr   2.1.4     v forcats 1.0.0
-- Conflicts ------------------------------------------ tidyverse_conflicts() --
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()


a. (_1 point_) What is your first guess *no calculating or simulating* - what is the chance that a person from the referenced population has COVID-19 given that they get a positive test result, assuming you know nothing about their symptoms or signs?


b. (_10 points_) Use the function described above to build your simulation to determine a numerical solution to part A - what is the chance that a person from the referenced population has COVID-19 given that they get a positive test result, assuming you know nothing about their symptoms or signs?

In [40]:
num_people = 100000
prop_covid = 5000 / num_people
pos_covid = 0.95
pos_nocovid = 0.05

prob_covid_given_positive <- function(num_people, prop_covid, pos_covid, pos_nocovid){
    list_result <- func_testing(num_people, prop_covid, pos_covid, pos_nocovid)
    
    answer1 <- list_result$prob_covid * list_result$prob_pos_given_covid /
        ((list_result$prob_pos_given_covid * list_result$prob_covid) 
         + (list_result$prob_pos_given_nocovid * list_result$prob_nocovid))
    
    return (answer1)
}

prob_covid_given_positive(num_people, prop_covid, pos_covid, pos_nocovid)

[1] 0.4969182

c. (_10 points_) This [clinical oncology article](https://www.clinicaloncology.com/COVID-19/Article/07-20/False-Negatives-Found-If-COVID-19-Testing-Done-Too-Soon/58781) references that for a particular test that the case false-negative rate was found to be about 40% when a test was done early on during symptoms. Simulate the chance that a random person has COVID-19 given that they get a negative test result. The false-positive rate is 10%, the false-negative rate is 40%, and the prevalence of COVID-19 in the population is 10,000/100,000.

In [41]:
num_people = 100000
prop_covid = 10000 / num_people
pos_covid = 0.6
pos_nocovid = 0.1

prob_covid_given_negative <- function(num_people, prop_covid, pos_covid, pos_nocovid){
    list_result <- func_testing(num_people, prop_covid, pos_covid, pos_nocovid)
    
    answer1 <- list_result$prob_covid * (1 - list_result$prob_pos_given_covid) / 
        (((1 - list_result$prob_pos_given_covid) * list_result$prob_covid) 
         + ((1 - list_result$prob_pos_given_nocovid) * list_result$prob_nocovid))
    
    return (answer1)
}

prob_covid_given_negative(num_people, prop_covid, pos_covid, pos_nocovid)


[1] 0.04814504